File: sort/HeapSort.cc
#include <iostream>

#include "utils.h"

using namespace std;

void BuildSort(IntArray &data, int index, int len)
{
    //置顶向下调整
    for (int i = index; i < data.size();)
    {
        int left_child_index = 2 * i + 1;
        int max_child_index = left_child_index;

        if (left_child_index >= len)
        {
            //无子节点
            break;
        }

        if (left_child_index + 1 < len && data[left_child_index + 1] > data[left_child_index])
        {
            max_child_index = left_child_index + 1;
        }

        if (data[i] > data[left_child_index])
        {
            //接下去的不用调整了
            break;
        }
        else
        {
            //交换
            int temp = data[max_child_index];
            data[max_child_index] = data[i];
            data[i] = temp;
        }
        i = max_child_index;
    }
}

void HeapSort(IntArray &data)
{
    for (int i = data.size() / 2 - 1; i >= 0; i--)
    {
        BuildSort(data, i, data.size());
    }

    for (int i = data.size(); i > 0; i--)
    {

        swap(data[0], data[i - 1]);
        BuildSort(data, 0, i - 1); //是i-1而不是i
    }
}

int main(int argc, char const *argv[])
{
    IntArray data;
    GetRandomArray(data, 0, 100, 100);
    cout << "Raw input:" << endl;

    PrintArray(data);

    HeapSort(data);

    cout << "Result output:" << endl;
    PrintArray(data);

    return 0;
}

File: sort/HeapSort.py
import utils

def build_heap(nums, heap_index, end):
    while heap_index <=end:

        left_index = 2 * heap_index + 1
        right_index = 2 * heap_index + 2
        max_child_index = left_index

        if left_index > end:
            return 

        if right_index <= end and nums[right_index] > nums[left_index]:
            max_child_index = right_index

        if nums[max_child_index] <= nums[heap_index]:
            return 

        nums[heap_index],nums[max_child_index] = nums[max_child_index],nums[heap_index]
        heap_index = max_child_index
        
def heap_sort(nums):
    for heap_index in range(int(len(nums)/2) - 1, -1, -1):
        build_heap(nums, heap_index, len(nums) - 1)

    for i in range(len(nums)-1, 0, -1):
        nums[i], nums[0] = nums[0], nums[i]
        build_heap(nums, 0, i-1)

def main():
    nums = utils.get_random(0,100,10)
    print("Raw input:")
    utils.print_nums(nums)
    heap_sort(nums)
    print("Result output:")
    utils.print_nums(nums)

if __name__ == "__main__":
    main()
